### PR TITLE
Fix integration tests

### PR DESCRIPTION
**Pull request by JulienMirval**
- Fix and expand integration tests
- Better error handling
- Better code reuse
- Increased scripts flexibility

### CODE REVIEW

File: src/lib/createTree.js [hunk 1]
```diff
@@ -3,60 +3,52 @@ const { v4: uuid } = require('uuid')
...
-  const querier = {
-    webhook: querierWebhooks.aggregationWebhook,
-    level: 0,
-    nbChild: 3,
-    aggregatorId: uuid(),
-    finalize: true
-  }
+const createTree = (treeStructure, nodesWebhooks) => {
+  let remainingNodes = nodesWebhooks
+  let matchingNodes = remainingNodes.slice()
```
Comment (paultranvan) [inline src/lib/createTree.js:12]: nit: In modern ES6 syntax, you can write `matchingNodes = [...remainingNodes]`

File: scripts/learning/dissecLearning.js [hunk 2]
```diff
@@ -19,27 +19,34 @@ const { createLogger } = require('../../src/targets/services/helpers/utils')
...
-  const aggregatorsWebhooks = aggregationNodes.filter(e => e.label !== uri)
-  const contributorsWebhooks = aggregatorsWebhooks
-
-  const contributors = createTree(
-    querierWebhooks,
-    aggregatorsWebhooks,
-    contributorsWebhooks
-  )
+
+  const treeStructure = [
```
Comment (paultranvan) [inline scripts/learning/dissecLearning.js]: Could you detail a bit in comment the tree structure and the expected params?

File: src/lib/createTree.js [hunk 3]
```diff
@@ -3,60 +3,52 @@ const { v4: uuid } = require('uuid')
...
+          remainingNodes.indexOf(matchingNodes.shift()),
+          1
+        )[0],
+        level: j,
+        nbChild:
+          j < treeStructure.length - 1 ? treeStructure[j + 1].numberOfNodes : 0,
+        parents: j > 0 ? lastLevel : undefined,
+        aggregatorId: uuid(),
+        finalize: j === 0
+      })
```
Comment (paultranvan) [inline src/lib/createTree.js]: That's kind of dark magic to me :sweat_smile: 
I'll trust you on this one, but it's a bit hard to read

File: src/lib/createTree.js [hunk 4]
```diff
@@ -3,60 +3,52 @@ const { v4: uuid } = require('uuid')
...
+  treeStructure.forEach((level, j) => {
+    const currentLevel = []
+    matchingNodes = remainingNodes.slice()
+
+    // Apply filters if there are any
+    if (level.mustInclude) {
+      if (level.mustInclude.length < level.numberOfNodes) {
+        throw new Error(
+          'Invalid tree structure: level not including enough nodes'
+        )
```
Comment (paultranvan) [inline src/lib/createTree.js:25]: I would rephrase the error in `Invalid tree structure: the "mustInclude" parameter must match the number of nodes`.

File: src/targets/services/categorize.js [hunk 5]
```diff
@@ -38,14 +38,21 @@ export const categorize = async () => {
...
-            new Date(e.date).valueOf() <=
+            new Date(e.date).valueOf() <
             new Date(filters.minOperationDate).valueOf()
         )
       : operations
 
     model = await Model.fromDocs(filteredOperations)
   }
 
+  console.log(
```
Comment (paultranvan) [inline src/targets/services/categorize.js]: It's easier for log readability to provide a bit of context, e.g. `console.log('Sum of occurences: ....')`

File: src/targets/services/contribution.js [hunk 6]
```diff
@@ -97,21 +97,27 @@ export const contribution = async () => {
   // Call webhooks of parents with the share
   for (let i in shareCodes) {
     // HACK: Using a delay to give enough time to the responding service to store shares
-    await new Promise(resolve => setTimeout(resolve, 5000))
+    await new Promise(resolve =>
+      setTimeout(resolve, 2000 * (1 + Math.random()))
+    )
```
Comment (paultranvan) [inline src/targets/services/contribution.js:102]: Why do you add random here?

File: src/targets/services/helpers/index.js [hunk 7]
```diff
@@ -1,9 +1,16 @@
+import largeClasses from '../../../assets/classes.json'
+import tinyClasses from '../../../assets/classesTiny.json'
+import largeVocabulary from '../../../assets/vocabulary.json'
+import tinyVocabulary from '../../../assets/vocabularyTiny.json'
```
Comment (paultranvan) [inline src/targets/services/helpers/index.js:4]: nit: it's better to use absolute paths, e.g `assets/vocabularyTiny.json`

File: src/targets/services/model.js [hunk 8]
```diff
@@ -62,28 +62,31 @@ export class Model {
       tokenizer,
       ...getClassifierOptions(this.uniqueY.length).initialization
     })
+    // Object.values(classes).map(category =>
+    //   classifier.initializeCategory(category)
+    // )
```
Comment (paultranvan) [inline src/targets/services/model.js:67]: To remove?

File: src/targets/services/model.js [hunk 9]
```diff
@@ -62,28 +62,31 @@ export class Model {
...
-          classifier.vocabulary[vocabulary[i]] = 0
         classifier.vocabulary[vocabulary[i]] += this.occurences[j][i]
-
-        if (!classifier.wordCount[this.uniqueY[j]])
-          classifier.wordCount[this.uniqueY[j]] = 0
         classifier.wordCount[this.uniqueY[j]] += this.occurences[j][i]
       }
     }
+
+    classifier.vocabularySize = vocabulary.length
```
Comment (paultranvan) [inline src/targets/services/model.js:89]: Does all of this really belong in the `feat: Flexible scripts ` commit ?

File: src/targets/services/contribution.js (hunk 6 above)
Comment (JulienMirval) [reply src/targets/services/contribution.js:102]: It's to reduce write conflicts

File: scripts/cli/categorize.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    client,
+    cutoffDate,
+    validationSet,
+    uri: options.supervisor,
+    pretrained: options.pretrained
+  })
+  log(`Accuracy = ${accuracy}`)
+}
+
+main()
```
[the remark this answers is not in the document]
Comment (JulienMirval) [reply scripts/cli/categorize.js:94]: There should not be any ambiguity because global categorization is not possible, only local categorization using a global model, which is already handled by the `categorize` script

File: src/targets/services/contribution.js (hunk 6 above)
Comment (paultranvan) [reply src/targets/services/contribution.js:103]: Might be worth commenting the why, it's not obvious ;)

File: scripts/learning/dissecLearning.js (hunk 2 above)
Comment (Dodecahedr0x) [reply scripts/learning/dissecLearning.js]: Addressed in @b4e99c84ebab9eecf66dcb10e65723691f572299

**Review by paultranvan — APPROVED** [review]
